In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
#importing dataset to a Pandas Dataframe
credit_card_data = pd.read_csv('creditcard.csv')

In [3]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [5]:
# Dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [7]:
# checking the number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [8]:
# distribution of legit transactions and fraudulent transactions
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

# 0 --> normal transactions
# 1 --> fraudulent transactions
# Hence, the data is highly unbalanced

In [14]:
# separating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data['Class'] == 1]

In [16]:
fraud.shape

(492, 31)

In [17]:
legit.shape

(284315, 31)

In [18]:
#statistical measures of the data
legit['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [19]:
fraud['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [20]:
# compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]


# Under-Sampling of data is observed
# We now build a Sample dataset containing similar distribution of normal transactions and fraudulent transactions, and in order to do this, we take 492 random values from the legit transactions and join it with the fraudulent dataset

In [21]:
legit_sample = legit.sample(n = 492)     #random sampling of legit data

In [22]:
# Concatenating two Dataframes
new_dataset = pd.concat([legit_sample, fraud], axis = 0)

In [23]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
198586  132553.0  0.039417  0.854441  0.262000 -0.608719  0.471513 -1.000020   
56754    47576.0  1.184447 -0.713744 -0.469846 -0.806087 -0.370748 -0.479665   
162762  115347.0  0.841180 -2.313056 -1.556714  0.319153  0.022429  1.989874   
230745  146458.0 -0.573725 -0.695654 -2.000683 -0.705536  2.071878 -0.509649   
218227  141225.0  2.067041 -1.906057 -0.939434 -1.471149 -1.551455 -0.428704   

              V7        V8        V9  ...       V21       V22       V23  \
198586  0.997741 -0.177479 -0.041627  ... -0.258979 -0.560527  0.068707   
56754   0.077937 -0.234042 -1.478882  ... -0.662079 -1.779111 -0.004216   
162762 -0.050673  0.470532  0.867001  ... -0.025381 -1.127727  0.030069   
230745  2.073066 -0.566677 -0.036187  ...  0.590846  2.053564  0.236075   
218227 -1.252706 -0.057399 -0.806038  ... -0.104488 -0.150869  0.196569   

             V24       V25       V26       V27       V28  Amount  Class  
198586 -0.090670 -0.494637  0.147762  0.250117  0.096185    3.59      0  
56754  -0.497630  0.147935  0.724375 -0.107567  0.009721  137.80      0  
162762 -0.949876 -0.977041  0.169861 -0.108396  0.026281  551.40      0  
230745 -0.939626 -1.345015  0.156211  0.052431  0.211263  226.00      0  
218227  0.603159 -0.346853 -0.212177 -0.011269 -0.021457  133.60      0  

[5 rows x 31 columns]

In [26]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [27]:
new_dataset['Class'].value_counts()

1    492
0    492
Name: Class, dtype: int64

In [28]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      90982.217480  0.092821  0.085458  0.122864 -0.027780  0.004627   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.028226  0.017892  0.044139  0.003215  ... -0.004521 -0.001767   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.036654 -0.003515  0.022882  0.058168 -0.029028  0.001355  0.003001   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       72.599553  
1      122.211321  

[2 rows x 30 columns]

# Splitting the data into Features and Targets

In [32]:
x = new_dataset.drop(columns = 'Class', axis = 1)
y = new_dataset['Class']

In [33]:
print(x)

            Time        V1        V2        V3        V4        V5        V6  \
198586  132553.0  0.039417  0.854441  0.262000 -0.608719  0.471513 -1.000020   
56754    47576.0  1.184447 -0.713744 -0.469846 -0.806087 -0.370748 -0.479665   
162762  115347.0  0.841180 -2.313056 -1.556714  0.319153  0.022429  1.989874   
230745  146458.0 -0.573725 -0.695654 -2.000683 -0.705536  2.071878 -0.509649   
218227  141225.0  2.067041 -1.906057 -0.939434 -1.471149 -1.551455 -0.428704   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [34]:
print(y)

198586    0
56754     0
162762    0
230745    0
218227    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


# Splitting the data into Training data and Test data:

In [35]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 2)

In [36]:
print(x.shape, x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(984, 30) (787, 30) (197, 30) (787,) (197,)


# Machine Learning Model Training
# Logistic Regression

In [40]:
model = LogisticRegression(max_iter = 1000)

In [41]:
#training the Logistic Regression Model with Training Data
model.fit(x_train, y_train)

LogisticRegression(max_iter=1000)

# Model Evaluation based on Accuracy Score

In [44]:
# accuracy on training data
x_train_prediction = model.predict(x_train)
training_data_accuracy = accuracy_score(x_train_prediction, y_train)

In [48]:
print("Accuracy on Training Data :", training_data_accuracy)

Accuracy on Training Data : 0.9466327827191868


In [49]:
# accuracy on training data
x_test_prediction = model.predict(x_test)
test_data_accuracy = accuracy_score(x_test_prediction, y_test)

In [51]:
print("Accuracy on Test Data :", test_data_accuracy)

Accuracy on Test Data : 0.934010152284264


# Hence, we see that the Training Data accuracy and Test Data accuracy scores are almost the same, which implies that our Credit Card Fraud Detection model is successful and has an accuracy of around 94%
